### PR TITLE
Make server keep alive

### PR DESCRIPTION
**Pull request by Nowilltolife**
- Added ability to switch between keep alive and non keep alive
- Add close callback
- Added ability to send data without recv first

### CODE REVIEW

File: libs/net/source/net/socket_server.cpp [hunk 1]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
-                if (!data.empty()) {
-                    std::vector<u8> result = callback(clientSocket, data);
-                    ::send(clientSocket, reinterpret_cast<char*>(result.data()), result.size(), 0);
-                    data.clear();
-                }
-                break;
-            } else {
-                std::copy(buffer.begin(), buffer.begin() + n, std::back_inserter(data));
+            int len = ::recv(clientSocket, reinterpret_cast<char*>(buffer.data()), buffer.size(), 0);
+            if(len > 0) {
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:94]: Missing space after if

File: libs/net/source/net/socket_server.cpp [hunk 2]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
-            if (n <= 0) {
-                if (!data.empty()) {
-                    std::vector<u8> result = callback(clientSocket, data);
-                    ::send(clientSocket, reinterpret_cast<char*>(result.data()), result.size(), 0);
-                    data.clear();
-                }
-                break;
-            } else {
-                std::copy(buffer.begin(), buffer.begin() + n, std::back_inserter(data));
+            int len = ::recv(clientSocket, reinterpret_cast<char*>(buffer.data()), buffer.size(), 0);
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:93]: I'd call this `receivedByteCount` or something

File: libs/net/source/net/socket_server.cpp [hunk 3]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
+            if(len > 0) {
+                std::copy(buffer.begin(), buffer.begin() + len, std::back_inserter(data));
+                continue;
             }
+
+            if(!data.empty()) {
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:99]: "

File: libs/net/source/net/socket_server.cpp [hunk 4]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
+
+                // callback with the data
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:101]: Capitalize first letter in comments

File: libs/net/source/net/socket_server.cpp [hunk 5]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
+                std::vector<u8> result = callback(clientSocket, data);
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:102]: Use `auto`

File: libs/net/source/net/socket_server.cpp [hunk 6]
```diff
@@ -78,28 +90,49 @@ namespace wolv::net {
...
+                    break;
+
+            }
+
+            // check if the client is still connected
+            if (len < 0 && errno == EAGAIN)
+                continue;
+
+            // if the client is not connected, break
+            close(clientSocket);
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:120]: `close()` doesn't exist on Windows. You only need `closeSocket`

File: libs/net/source/net/socket_server.cpp [hunk 7]
```diff
@@ -68,8 +68,20 @@ namespace wolv::net {
...
-        std::vector<u8> buffer(this->m_bufferSize);
+    void SocketServer::send(SocketHandle socket, const std::vector<u8> &data) const {
+        ::send(socket, reinterpret_cast<const char*>(data.data()), data.size(), 0);
+    }
+
+    void SocketServer::send(SocketHandle socket, const std::string &data) const {
+        ::send(socket, data.c_str(), data.size(), 0);
+    }
+
+    void SocketServer::close(wolv::net::SocketHandle socket) const {
```
Comment (WerWolv) [inline libs/net/source/net/socket_server.cpp:79]: Why is this function needed?

**Review by WerWolv — CHANGES_REQUESTED** [review]

File: libs/net/source/net/socket_server.cpp (hunk 7 above)
Comment (Nowilltolife) [reply libs/net/source/net/socket_server.cpp:79]: Forcibly closing a keep alive client, i added this because i thought closeSocket was private

**Review by WerWolv — APPROVED** [review]

**Pull request merged**